### PR TITLE
Enable Creating Events & Todo on Teacher Calendars

### PR DESCRIPTION
**Pull request by suhaibabsi-inst**
refs: MBL-17889
affects: Teacher
release note: Enabling events & todos creation on Teacher's calendars

## Test Plan
Done smoke tests for both "Add Todo" & "Add Event" actions, the methodology:

- Logged-in to Teacher app with user with `teacher` role assigned to him of at least one course.
- Navigated to "Calendar" tab, tapped on Calendars to view all calendars assigned to the user.
- On the top right corner, tapped on "+" button, selected any of those actions.
- Filled in form and hit submit.
- On both forms, tried to create instances on all available Calendars, and they seem to work as expected.

## Screen Recording

https://github.com/user-attachments/assets/a77aee4b-bef4-4806-9332-1d98b0d9993f

## Checklist

- [ ] Follow-up e2e test ticket created
- [ ] A11y checked
- [x] Tested on phone
- [x] Tested on tablet
- [x] Tested in dark mode
- [x] Tested in light mode
- [ ] Approve from product

### CODE REVIEW

**Comment by inst-danger** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>Build failed, skipping coverage check</td>
    </tr>
  
<tr>
      <td>:no_entry_sign:</td>
      <td>

  Please run `yarn update-headers`. The following do not have the correct license header: 
* Core/Core/Planner/CalendarFilter/Model/UseCase/GetTeacherCalendarFilters.swift
  </td>
    </tr>
  </tbody>
</table>



#### Release Note: 
Enabling events & todos creation on Teacher's calendars

#### Affected Apps: Teacher

[MBL-17889](https://instructure.atlassian.net/browse/MBL-17889)
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 80640e5feeeea8cc6a3b7d362bdf8a978bd61b62
</p>

**Comment by inst-danger** [comment]
<details><summary>Teacher Build QR Code:</summary>

</details>

**Review by rh12 — DISMISSED** [review]
QA + 1

My understanding after review is that the `Purpose` enum governs whether we fetch courses via `GetCurrentUserCoursesRequest` or `GetCoursesRequest` (with `enrollmentType: .teacher`), but only for teacher app.
And the `.creating` case is used only for event creation/update but not for todo creation/update (or calendar display).
Could you please elaborate a bit on why is this needed? (Either here or in PR description or in code comment.)

File: Core/Core/Planner/CalendarFilter/Model/CoreData/CDCalendarFilterEntry.swift [hunk 1]
```diff
@@ -19,11 +19,26 @@
 import CoreData
 import SwiftUI
 
+public enum CDCalendarFilterPurpose: Int16 {
+    case viewing = 1
+    case creating = 2
+    case unknown = 0
```
[the remark this answers is not in the document]
Comment (suhaibabsi-inst) [reply Core/Core/Planner/CalendarFilter/Model/CoreData/CDCalendarFilterEntry.swift:25]: @rh12 This distinction is actually needed for `scope` of `GetCalendarFilters`. We store this value to CoreData database, so the `reset` function remove relevant objects per use case. So, this value is used in `CalendarFilters` fetching use case. 

To limit things when using the use case for teachers, internal `Purpose` was defined on `GetTeacherCalendarFilters` so to force the caller be aware when use this type either to view or create instances on Teacher calendar.

**Comment by suhaibabsi-inst** [comment]
> QA + 1
> 
> My understanding after review is that the `Purpose` enum governs whether we fetch courses via `GetCurrentUserCoursesRequest` or `GetCoursesRequest` (with `enrollmentType: .teacher`), but only for teacher app. And the `.creating` case is used only for event creation/update but not for todo creation/update (or calendar display). Could you please elaborate a bit on why is this needed? (Either here or in PR description or in code comment.)

The need to determine the context at which we fetch calendars for Teacher was dictated by the API. So, if you try to call create event API on a calendar where you have no teacher role on it, it will respond with an error upon that. While this is not the same for Todos, you can create those on all calendars associated with you regardless of your role.

File: Core/Core/Planner/CalendarFilter/Model/CoreData/CDCalendarFilterEntry.swift (hunk 1 above)
[the remark this answers is not in the document]
Comment (suhaibabsi-inst) [reply Core/Core/Planner/CalendarFilter/Model/CoreData/CDCalendarFilterEntry.swift:25]: Another reason why I went on defining another type on the Teacher use case to represent the purpose, is that we don't yet whether that would be required for other cases or not. And would rather keep other use cases unaware of it as long as they don't need it.  You can see `GetParentCalendarFilters` as an example for that.

[the remark this answers is not in the document]
Comment (suhaibabsi-inst) [reply Core/Core/Planner/CalendarFilter/Model/CoreData/CDCalendarFilterEntry.swift:25]: For the use of `String` as the raw value of `CDCalendarFilterPurpose` and the use of `unknown` case. I should say that is a matter of preference to me, to use `Int16` to express enum cases in CoreData base. 

From experience, cases where you're using enums to decide on a some strategy, using optional or String as raw value can lead a situation where you need to handle this ambiguity  in many places in code, which get could things out of control or at least adds a lot of mental load trying to maintain that. That's why I found this technique more definitive for me on most cases.

Let me know your thoughts about it, I mean why do you think using `String` rawValue and `Optional` value for `unknown` case is better technique than this one?

[the remark this answers is not in the document]
Comment (rh12) [reply Core/Core/Planner/CalendarFilter/Model/CoreData/CDCalendarFilterEntry.swift:25]: At a glance, it seemed more simple that way, but I don't mind it as it is right now. Thanks for elaborating.

**Review by rh12 — APPROVED** [review]

**Review by vargaat — APPROVED** [review]
QA+1

**Pull request merged**